test/lit/passes/optimize-casts.wast

```diff
@@ -387,6 +420,789 @@
+        (local.get $x)
+      )
+    )
+    (drop
+      (local.get $x)
+    )
+  )
+
+  ;; CHECK:      (func $move-cast-6 (type $ref|struct|_=>_none) (param $x (ref struct))
+  ;; CHECK-NEXT:  (local $1 (ref $B))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.tee $1
+  ;; CHECK-NEXT:    (ref.cast $B
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $move-cast-6 (param $x (ref struct))
+    (drop
+      ;; This is already the most refined cast, so nothing will be moved.
+      (ref.cast $B
+        (local.get $x)
+      )
+    )
+    (drop
+      (local.get $x)
+    )
+    (drop
+      ;; Converted to $B by ReFinalize().
+      (ref.cast $A
+        (local.get $x)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $move-already-refined-local (type $ref|$B|_=>_none) (param $x (ref $B))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $x)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $move-already-refined-local (param $x (ref $B))
+    (drop
+      (local.get $x)
+    )
+    (drop
+      ;; Since we know $x is of type $B, this cast to a less refined type $A
+      ;; will not be moved higher.
+      (ref.cast $A
+        (local.get $x)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $avoid-erroneous-cast-move (type $none_=>_none)
+  ;; CHECK-NEXT:  (local $a (ref null $A))
+  ;; CHECK-NEXT:  (local $b (ref $A))
+  ;; CHECK-NEXT:  (local $2 (ref $A))
+  ;; CHECK-NEXT:  (local.set $b
+  ;; CHECK-NEXT:   (local.tee $2
+  ;; CHECK-NEXT:    (ref.as_non_null
+  ;; CHECK-NEXT:     (local.get $a)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $D
+  ;; CHECK-NEXT:    (ref.as_non_null
+  ;; CHECK-NEXT:     (local.get $2)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $avoid-erroneous-cast-move
+    ;; This tests whether we are able to avoid moving a cast earlier if
+    ;; we know doing so will violate typing rules.
+    (local $a (ref null $A))
+    (local $b (ref $A))
+    (local.set $b
+      (ref.as_non_null
+        ;; We could move the ref.cast $D here. However, as $a is already known
+        ;; to have type ref null $A, not type $D, it would fail. Moving the cast
+        ;; will also cause the local.set $b to fail, since $b is of type
+        ;; ref null $A, not $D.
+
+        ;; By checking the type of $a, we can avoid moving the ref.cast $D here.
+        ;; If the type of $a is wrong to begin with, the error will be caught
+        ;; before the OptimizeCast pass is run.
```

Comment:
```suggestion
```
This comment duplicates the above, I think. And no need to mention the last sentence - that's just the normal situation that we can assume modules are valid (in passes, since we validate before).